File: android/src/main/kotlin/com/cloudwebrtc/webrtc/controller/RtpTransceiverController.kt
package com.cloudwebrtc.webrtc.controller

import com.cloudwebrtc.webrtc.model.RtpTransceiverDirection
import com.cloudwebrtc.webrtc.proxy.MediaStreamTrackProxy
import com.cloudwebrtc.webrtc.proxy.RtpTransceiverProxy
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/**
 * Controller for the [RtpTransceiverProxy] functional.
 *
 * @property binaryMessenger messenger used for creating new [MethodChannel]s.
 * @property transceiver underlying [RtpTransceiverProxy] on which method calls will be performed.
 */
class RtpTransceiverController(
    private val binaryMessenger: BinaryMessenger,
    private val transceiver: RtpTransceiverProxy
) : MethodChannel.MethodCallHandler, IdentifiableController {
    /**
     * Unique ID of the [MethodChannel] of this controller.
     */
    private val channelId = nextChannelId()

    /**
     * Channel which will be listened for the [MethodCall]s.
     */
    private val methodChannel =
        MethodChannel(binaryMessenger, ChannelNameGenerator.withId("RtpTransceiver", channelId))

    init {
        methodChannel.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "setDirection" -> {
                val direction = RtpTransceiverDirection.fromInt(call.argument("direction")!!)
                transceiver.setDirection(direction)
                result.success(null)
            }
            "getMid" -> {
                result.success(transceiver.getMid())
            }
            "getDirection" -> {
                result.success(transceiver.getDirection().value)
            }
            "getCurrentDirection" -> {
                result.success(transceiver.getCurrentDirection()?.value)
            }
            "getSender" -> {
                result.success(
                    RtpSenderController(
                        binaryMessenger,
                        transceiver.getSender()
                    ).asFlutterResult()
                )
            }
            "stop" -> {
                transceiver.stop()
                result.success(null)
            }
            "dispose" -> {
                dispose()
                result.success(null)
            }
        }
    }

    /**
     * Converts this [RtpTransceiverController] to the Flutter's method call result.
     *
     * @return [Map] generated from this controller which can be returned to the Flutter side.
     */
    fun asFlutterResult(): Map<String, Any?> {
        return mapOf(
            "channelId" to channelId,
            "sender" to RtpSenderController(
                binaryMessenger,
                transceiver.getSender()
            ).asFlutterResult(),
            "mid" to transceiver.getMid() as Any?
        )
    }

    /**
     * Closes method channel of this [RtpTransceiverController].
     */
    private fun dispose() {
        methodChannel.setMethodCallHandler(null)
    }
}

File: ios/Classes/controller/RtpTransceiverController.swift
import Flutter

/// Controller for the `RtpTransceiver`.
class RtpTransceiverController {
  /// Flutter messenger for creating channels.
  private var messenger: FlutterBinaryMessenger

  /// Instance of `RtpTransceiver` proxy.
  private var transceiver: RtpTransceiverProxy

  /// ID of channels created for this controller.
  private var channelId: Int = ChannelNameGenerator.nextId()

  /// Method channel for communicating with Flutter side.
  private var channel: FlutterMethodChannel

  /// Creates new `RtpTransceiverController` for the provided `RtpTransceiverProxy`.
  init(messenger: FlutterBinaryMessenger, transceiver: RtpTransceiverProxy) {
    let channelName = ChannelNameGenerator.name(name: "RtpTransceiver", id: self.channelId)
    self.messenger = messenger
    self.transceiver = transceiver
    self.channel = FlutterMethodChannel(name: channelName, binaryMessenger: messenger)
    self.channel.setMethodCallHandler({ (call, result) in
      try! self.onMethodCall(call: call, result: result)
    })
  }

  /// Handles all supported Flutter method calls for the `RtpTransceiverProxy`.
  func onMethodCall(call: FlutterMethodCall, result: FlutterResult) throws {
    let argsMap = call.arguments as? [String: Any]
    switch call.method {
    case "setDirection":
      let direction = argsMap!["direction"] as? Int
      self.transceiver.setDirection(direction: TransceiverDirection(rawValue: direction!)!)
      result(nil)
    case "setRecv":
      let enabled = argsMap!["enabled"] as? Bool
      self.transceiver.setRecv(recv: enabled!)
      result(nil)
    case "setSend":
      let enabled = argsMap!["enabled"] as? Bool
      self.transceiver.setSend(send: enabled!)
      result(nil)
    case "getMid":
      let mid = self.transceiver.getMid()
      result(mid)
    case "getDirection":
      let direction = self.transceiver.getDirection()
      result(direction.rawValue)
    case "stop":
      result(nil)
    case "dispose":
      self.channel.setMethodCallHandler(nil)
      result(nil)
    default:
      result(FlutterMethodNotImplemented)
    }
  }

  /// Converts this controller to the Flutter method call response.
  func asFlutterResult() -> [String: Any] {
    return [
      "channelId": self.channelId,
      "sender": RtpSenderController(messenger: self.messenger, rtpSender: transceiver.getSender())
        .asFlutterResult(),
      "mid": transceiver.getMid(),
    ]
  }
}
